core/modules/system/system.admin.inc

```diff
@@ -1000,16 +999,21 @@ function system_modules_submit($form, &$form_state) {
   }
 
   // Check if a previous search is present and preserve it on redirect.
-  if (!empty($form_state['values']['search'])) {
+  if (!empty($form_state['values']['search']) && sizeof($actions['enable']) != 0)) {
     $parameter = $form_state['values']['search'];
     $form_state['redirect'] = array(
       'admin/modules',
       array('query' => array('search' => $parameter))
     );
   }
-  else {
+  elseif(sizeof($actions['enable']) != 0) {
     $form_state['redirect'] = 'admin/modules';
   }
+  else {
```

Comment:
I find repetition of the `sizeof($actions['enable']) != 0` condition in 2 separate conditionals superfluous. I think that this is a cleaner way to do it:

```
if (sizeof($actions['enable']) != 0) {
  // If there was a search query, preserve it when redirecting...
  if (!empty($form_state['values']['search'])) {
    $parameter = $form_state['values']['search'];
    $form_state['redirect'] = array('admin/modules', array('query' => array('search' => $parameter)));
  }
  // ...otherwise simply redirect to the modules list page.
  else {
    $form_state['redirect'] = 'admin/modules';
  }
}
else {
  $form_state['redirect'] = 'admin/modules/uninstall';
}
```

This is purely from looking at the code from a novice dev PoV; this may need to be rethought, considering the logic we are introducing.